### ISSUE TITLE
URLconfig doesn't allow delayed `setup(self)`

### ISSUE THREAD

**Issue by JoranAngevaare** [XENONnT/straxen#803]
For plugins, we use the `setup(self)` method to do heavy initialization (like downloading tensorflow files). This allows to quickly initialize plugins-classes, but only require the setup argument to be called when we actually need to compute the data from that plugin (see [this line](https://github.com/AxFoundation/strax/blob/39890e94de17b5b508c3fa634123a908d55908f0/strax/context.py#L406)). With URL configs, this is now moved to the [`_set_plugin_config`](https://github.com/AxFoundation/strax/blob/39890e94de17b5b508c3fa634123a908d55908f0/strax/context.py#L409) method of the context. 

Now, if we do a scan of what data is stored, this requires setting the configs (but we don't need a setup!). We just [need the plugin with it's config according to the config of the context](https://github.com/AxFoundation/strax/blob/39890e94de17b5b508c3fa634123a908d55908f0/strax/context.py#L574). This would be slowed down significantly if we download neural nets etc.

One example is [the tensorflow model](https://github.com/XENONnT/straxen/blob/20dea986790ef168ba7052d652a7aa19ab836943/straxen/plugins/position_reconstruction.py#L74), but we have many more.

**Comment by jmosbacher** [comment]
Im not sure i follow. Why would the expensive operation be done in _set_plugin_config? the lookup is only performed when the value is accessed from the plugin. so in the tensorflow example only when you try to access `self.model` would the model be loaded. the value inside plugin.config dictionary is always the value set by the user.

**Comment by jmosbacher** [comment]
Just the opposite, the URLConfig usually removes the need for a setup(self) method because the construction of the actual value is only done when the config is actually used (usually in the compute method) and you can simply enable the cache if its an expensive operation so next time the attribute is accessed it will not need the expensive operation.

**Comment by jmosbacher** [comment]
One more advantage I can think of in using the URLConfig cache is that its cache is shared among all instances of the same class in every python process, so in a multi-threaded situation you are only doing the computation once for all threads in the same process.

**Comment by JoranAngevaare** [comment]
Okay, thanks Yossi for explaining and quite I'm happy I'm wrong in that case. 

But then it's important to cache e.g. the CMT values to avoid doing the same query for each chunk if I understand you correctly?

**Comment by jmosbacher** [comment]
I thought CMT caches its own values on a global level no?

**Comment by jmosbacher** [comment]
I think in most cases we can implement the caching inside the protocol itself, like is done in CMT and the get_resource method. The URLConfig internal caching is meant more for very complex cases e.g. a chained protocol where one or more of the values being passed between protocols is not hashable so cant be cached. in such a case the internal cache would be useful because it caches the final result based on the entire URL which is always hashable.
